### PR TITLE
Feature/news api

### PR DESCRIPTION
**Pull request by Samy-33**
Added News Feature using api of newsapi.org  

- Can get # of articles from given source and in provided order (latest, popular etc)  
- Can get list of sources for given **language, country and category**  

## Screenshots:  

Waiting for review.

### CODE REVIEW

File: bot_chucky/bot.py [hunk 1]
```diff
@@ -172,3 +175,41 @@ def send_stack_questions(self, id_, **kwargs):
...
+	
+    def get_article(self, id_:str, source, count, order=None):
+        """
+        ## This function gets #count articles from 'source' sorted by 'order'
+        :param id_: facebook user id
+        :param source: Source of news article
+        :param: count: Number of articles to be sent
+        :order: latest, top, or popular
+        """
+        if(self.news.get_key() is None):
```
Comment (MichaelYusko) [inline bot_chucky/bot.py]: We don't need parentheses here;)

File: bot_chucky/bot.py [hunk 2]
```diff
@@ -172,3 +175,41 @@ def send_stack_questions(self, id_, **kwargs):
...
+            raise BotChuckyError("Articles are available only with the newsapi.org key")
+            
+        data = self.news.get_article(source, min(count, 10), order)
```
Comment (MichaelYusko) [inline bot_chucky/bot.py]: We can do here, more readable?

File: bot_chucky/bot.py [hunk 3]
```diff
@@ -172,3 +175,41 @@ def send_stack_questions(self, id_, **kwargs):
...
+        
+        for article in data:
+            message = f"\nBy {article['author']}\nTitle:{article['title']}\n"
+            message += f"Desc: {article['description']}\nRead more: {article['url']}"
+            print(self.send_message(id_, message[:min(len(message), 600)]))
```
Comment (MichaelYusko) [inline bot_chucky/bot.py]: The same here and why we need to use `print` function here?

File: bot_chucky/bot.py [hunk 4]
```diff
@@ -172,3 +175,41 @@ def send_stack_questions(self, id_, **kwargs):
...
+    
+    def get_sources_list(self, id_:str, count, category=None, language=None, country=None):
+        """
+        :param id_: facebook user id
+        :param count: # of sources user wants to list
+        :param category: #Main category in which the source posts articles
+        :language: Understandable -> en, de or fr
+        :country: Country in which the source belongs
+        """
+        data = self.news.get_sources(min(count, 20), category, language, country)
```
Comment (MichaelYusko) [inline bot_chucky/bot.py]: We can do here, more readable?

File: bot_chucky/helpers.py [hunk 5]
```diff
@@ -182,6 +182,130 @@ def search(self, artist=None):
...
+        """
+        Fetches the mapping between the source codes and source names, according
+        to the params given
+        """
+        from .constants import NEWS_URL
+        if (category is not None and category not in ['bsns', 'entnt', 'gmg',\
+                                                      'gnrl', 'music', 'pltcs',\
+                                                      'scntr', 'sport', 'tech']):
+            raise BotChuckyError(f'Invalid category, choose from {self.categories}')
+        elif (category is not None):
```
Comment (MichaelYusko) [inline bot_chucky/helpers.py]: We don't need parentheses here;)

File: bot_chucky/helpers.py [hunk 6]
```diff
@@ -182,6 +182,130 @@ def search(self, artist=None):
...
+            category = f'category={self.categories[category]}'
+        else:
+            category = ''
+        
+        
+        if (language is not None and language not in ['en', 'de', 'fr']):
```
Comment (MichaelYusko) [inline bot_chucky/helpers.py]: The same here

File: bot_chucky/helpers.py [hunk 7]
```diff
@@ -182,6 +182,130 @@ def search(self, artist=None):
...
+            raise BotChuckyError(f'Invalid language, choose from {self.languages}')
+        elif (language is not None):
```
Comment (MichaelYusko) [inline bot_chucky/helpers.py]: The same

File: bot_chucky/bot.py (hunk 4 above)
Comment (Samy-33) [inline bot_chucky/bot.py]: Ok, thanks. I will do it. :+1:

File: bot_chucky/helpers.py [hunk 8]
```diff
@@ -182,6 +182,130 @@ def search(self, artist=None):
...
+            language= f'&language={language}'
+        else:
+            language= ''
+        
+        
+        if (country is not None and country not in ['au', 'de', 'gb', 'in', 'us']):
```
Comment (MichaelYusko) [inline bot_chucky/helpers.py]: the same for this line;)

File: bot_chucky/helpers.py [hunk 9]
```diff
@@ -182,6 +182,130 @@ def search(self, artist=None):
...
+            raise BotChuckyError(f'Invalid country, choose from {self.countries}')
+        elif (country is not None):
```
Comment (MichaelYusko) [inline bot_chucky/helpers.py]: and for this line

File: bot_chucky/helpers.py [hunk 10]
```diff
@@ -182,6 +182,130 @@ def search(self, artist=None):
...
+            country = f'&country={country}'
+        else:
+            country = ''
+        
+        url = f'{NEWS_URL}sources?{category}{language}{country}'
+        print(url)
+        data = r.get(url).json()
+        
+        if(len(data['sources'])==0):
```
Comment (MichaelYusko) [inline bot_chucky/helpers.py]: The same here

File: examples/flask_example.py [hunk 11]
```diff
@@ -44,6 +44,16 @@ def handle_messages():
 
                     # Send stackoverflow questions
                     bot.send_stack_questions(sender_id, title=text)
+                    
+                    # Send List of News Sources
+                    if(text[:9].lower()=='news list'):
```
Comment (MichaelYusko) [inline examples/flask_example.py]: The same here

File: examples/flask_example.py [hunk 12]
```diff
@@ -44,6 +44,16 @@ def handle_messages():
...
+                        bot.get_sources_list(sender_id, count=10, country="in")
+                    
+                    #Send 5 latest news articles from source
+                    if(text[:4].lower() == 'news'):
```
Comment (MichaelYusko) [inline examples/flask_example.py]: The same here.

**Comment by MichaelYusko** [comment]
@Samy-33 Hi, reviewed your a PR, thanks! nice one! but I left a couple of comment;)

File: examples/flask_example.py (hunk 12 above)
Comment (Samy-33) [inline examples/flask_example.py]: Got it, the "parentheses" ;)

**Comment by Samy-33** [comment]
Yes, removed unnecessary parentheses and made is more readable. :)

**Comment by MichaelYusko** [comment]
@Samy-33 Thanks;) I'll check in 2-3 hours

**Comment by MichaelYusko** [comment]
@Samy-33 Merged;) add your name into `CONTRIBUTORS.md` file

**Pull request merged**